File: Data Structures & Algorithms/combination-target-sum/submission-0.cpp
class Solution {
public:
    vector<vector<int>> v;
    vector<int> ans;
    void dfs(vector<int> nums, int index, int target) {
        if (index >= nums.size() || target <0)
            return;
        if (target == 0) {
            v.push_back(ans);
            return;
        }
        ans.push_back(nums[index]);
        dfs(nums, index, target-nums[index]);
        ans.pop_back();
        dfs(nums, index+1, target);
    }
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        dfs(nums, 0, target);
        return v;
    }

};

File: Data Structures & Algorithms/permutations/submission-0.py
class Solution:
    
    def permute(self, nums: List[int]) -> List[List[int]]:
        seen = set()
        ans = []
        def dfs(i, v):
            if len(v) == len(nums):
                ans.append(v.copy())
                return
            for j in range(0, len(nums)):
                if j not in seen:
                    seen.add(j)
                    v.append(nums[j])
                    dfs(j+1, v) 
                    v.pop()
                    seen.remove(j)
        dfs(0, [])
        return ans